### ISSUE TITLE
Train Once with a Weighted Loss without Training Group Information

### ISSUE THREAD

**Issue by LinkToPast1900** [anniesch/jtt#3]
Hi, @anniesch , thanks for your great work. 

Using weighted loss seems to be a common strategy that forces the model to focus on hard examples. 
So, I am wondering whether the weighted loss strategy should be added as a simple baseline besides CVaR DRO and LfF?

Considering the following weighted loss, which down-weights those predicted correctly examples directly, on the MultiNLI task

```
class CustomLoss(nn.modules.loss._WeightedLoss):
    def __init__(self, weight=None):
        super(CustomLoss, self).__init__()
        self.weight = weight # weight parameter will act as the alpha parameter to balance class weights
        
    def forward(self, input, target):
        ce_loss = F.cross_entropy(input, target, reduction="none", weight=self.weight)
        
        prediction = torch.argmax(input, dim=-1)
        alpha = 1-((prediction==target) * 0.75)
        
        loss = alpha * ce_loss

        return loss 
```

first, I change https://github.com/anniesch/jtt/blob/72a9e2b063070523cb54f820056952e5c4c1fb02/run_expt.py#L166
```
#         criterion = torch.nn.CrossEntropyLoss(reduction="none")
        criterion = CustomLoss()
```
and then run
```
CUDA_VISIBLE_DEVICES=1 python run_expt.py -s confounder -d MultiNLI -t gold_label_random -c sentence2_has_negation --batch_size 32 --root_dir ./ --n_epochs 5 --aug_col None --log_dir results/MultiNLI/MultiNLI_sample_exp/ERM_upweight_0_epochs_5_lr_2e-05_weight_decay_0.0/model_outputs --metadata_path results/MultiNLI/MultiNLI_sample_exp/metadata_aug.csv --lr 2e-05 --weight_decay 0.0 --up_weight 0 --metadata_csv_name metadata.csv --model bert --use_bert_params 1 --loss_type erm
```

and obtain **Avg Acc. 82.15%** and **Worst-group Acc. 71.86%** at epoch 4 (selected based on the highest worst-group accuracy on validation set).
![image](https://user-images.githubusercontent.com/85857950/131210112-66848804-9f54-4826-84b2-0dc97a4bde87.png)
which seems to be comparable with CVaR DRO and LfF (did I miss something here?)
![image](https://user-images.githubusercontent.com/85857950/131210313-b8499e9a-19dd-47ce-9050-1748979a73e4.png)

**Comment by anniesch** [comment]
Hi @LinkToPast1900, thanks so much for this comment! The approach of using a weighted loss could certainly be added as a baseline, and it would be interesting to explore how such an approach would compare to JTT and other methods. It does look like this method compares favorably to CVaR DRO and LfF here on MultiNLI. One key difference between this approach and JTT is that here the weights are recomputed each step, which gives it a similar flavor to CVaR DRO, where the error set to be upweighted is recomputed each step. We found in our analysis that not dynamically recomputing the error set at each step was a key factor in the effectiveness of JTT. More exploration in this area, however, such as with your proposed approach, would be very interesting and could lead to new insights.

**Comment by LinkToPast1900** [comment]
Thanks
